File: gestaltd/internal/daemon/secrets.go
package daemon

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"
	"unicode/utf8"

	"golang.org/x/term"

	"github.com/valon-technologies/gestalt/server/internal/config"
)

const (
	managedSecretsAdminPath  = "/admin/api/v1/managed-secrets"
	managedSecretMaxBytes    = 64 * 1024
	managedSecretHTTPTimeout = 30 * time.Second
)

type managedSecretClient struct {
	BaseURL    string
	Token      string
	HTTPClient *http.Client
}

func runSecrets(args []string) error {
	if len(args) == 0 {
		printSecretsUsage(os.Stderr)
		return flag.ErrHelp
	}

	switch args[0] {
	case "-h", "--help", "help":
		printSecretsUsage(os.Stdout)
		return nil
	case "create":
		return runManagedSecretWrite(args[1:], "create")
	case "rotate":
		return runManagedSecretWrite(args[1:], "rotate")
	case "list":
		return runManagedSecretList(args[1:])
	case "describe":
		return runManagedSecretDescribe(args[1:])
	case "retire":
		return runManagedSecretRetire(args[1:])
	default:
		return fmt.Errorf("unknown secrets command %q", args[0])
	}
}

func printSecretsUsage(w io.Writer) {
	writeUsageLine(w, "Usage:")
	writeUsageLine(w, "  gestaltd secrets <command> [flags]")
	writeUsageLine(w, "")
	writeUsageLine(w, "Commands:")
	writeUsageLine(w, "  create     Create a new managed secret version; fails if the name already exists")
	writeUsageLine(w, "  rotate     Rotate an existing managed secret by creating a new version")
	writeUsageLine(w, "  list       List managed secret metadata; values are never returned")
	writeUsageLine(w, "  describe   Show one secret's metadata and recent audit history")
	writeUsageLine(w, "  retire     Mark a secret retired while retaining history")
	writeUsageLine(w, "")
	writeUsageLine(w, "Commands use GESTALT_URL/GESTALT_API_KEY or credentials from `gestalt init` and `gestalt auth login`.")
	writeUsageLine(w, "Values are never accepted as command-line arguments.")
}

type managedSecretWriteArgs struct {
	name     string
	ownerApp string
	reason   string
}

func runManagedSecretWrite(args []string, operation string) error {
	fs := flag.NewFlagSet("gestaltd secrets "+operation, flag.ContinueOnError)
	fs.Usage = func() { printManagedSecretWriteUsage(fs.Output(), operation) }
	opts := managedSecretWriteArgs{}
	fs.StringVar(&opts.ownerApp, "owner-app", "", "app that owns this secret")
	fs.StringVar(&opts.reason, "reason", "", "why this value is being written")
	if err := parseInterspersed(fs, args); err != nil {
		return err
	}
	positionals := fs.Args()
	if len(positionals) != 1 {
		return fmt.Errorf("secret name is required")
	}
	opts.name = positionals[0]
	if strings.TrimSpace(opts.reason) == "" {
		return fmt.Errorf("--reason is required")
	}
	value, err := readManagedSecretValue(opts)
	if err != nil {
		return err
	}

	body := map[string]any{
		"name":      opts.name,
		"operation": operation,
		"ownerApp":  opts.ownerApp,
		"reason":    opts.reason,
		"source":    "gestaltd-cli",
	}
	if value == "" {
		body["generate"] = true
	} else {
		body["value"] = value
	}

	client, err := newManagedSecretClient()
	if err != nil {
		return err
	}
	var response managedSecretWriteResponse
	if err := client.post(context.Background(), managedSecretsAdminPath, body, &response); err != nil {
		return fmt.Errorf("failed to write managed secret: %w", err)
	}
	printManagedSecretWriteResponse(os.Stdout, response)
	return nil
}

func printManagedSecretWriteResponse(w io.Writer, response managedSecretWriteResponse) {
	_, _ = fmt.Fprintf(w, "Secret: %s\n", response.Secret.Name)
	if response.Secret.OwnerApp != "" {
		_, _ = fmt.Fprintf(w, "Owner app: %s\n", response.Secret.OwnerApp)
	}
	_, _ = fmt.Fprintf(w, "Version: %d\n", response.Version.Version)
	_, _ = fmt.Fprintf(w, "Created at: %s\n", response.Version.CreatedAt)
	if response.GeneratedValue != "" {
		_, _ = fmt.Fprintln(w, "Generated value (shown once):")
		_, _ = fmt.Fprintln(w, response.GeneratedValue)
	}
	if response.RolloutRequired {
		_, _ = fmt.Fprintln(w, "Rollout required: yes")
		_, _ = fmt.Fprintln(w, "Gestalt resolves this secret during startup; roll out the runtime to load the new version.")
	} else {
		_, _ = fmt.Fprintln(w, "Rollout required: no")
	}
}

func printManagedSecretWriteUsage(w io.Writer, operation string) {
	writeUsageLine(w, "Usage:")
	writeUsageLine(w, "  gestaltd secrets "+operation+" NAME --reason TEXT")
	writeUsageLine(w, "")
	writeUsageLine(w, "A value is read from piped stdin or a hidden terminal prompt.")
	writeUsageLine(w, "With no value, the server generates a secret and prints it once.")
	writeUsageLine(w, "Values are never accepted as command-line arguments.")
	writeUsageLine(w, "")
	writeUsageLine(w, "Flags:")
	writeUsageLine(w, "  --owner-app     App that owns this app-scoped secret")
	writeUsageLine(w, "  --reason        Why this value is being written (required)")
}

func readManagedSecretValue(opts managedSecretWriteArgs) (string, error) {
	if term.IsTerminal(int(os.Stdin.Fd())) {
		first, err := readManagedSecretPassword("Secret value: ")
		if err != nil {
			return "", err
		}
		second, err := readManagedSecretPassword("Confirm secret value: ")
		if err != nil {
			return "", err
		}
		if first != second {
			return "", errors.New("secret values do not match")
		}
		return validateManagedSecretValue([]byte(first))
	}
	raw, err := io.ReadAll(os.Stdin)
	if err != nil {
		return "", fmt.Errorf("failed to read secret value from stdin: %w", err)
	}
	return validateManagedSecretValue(raw)
}

func validateManagedSecretValue(raw []byte) (string, error) {
	if len(raw) > managedSecretMaxBytes {
		return "", fmt.Errorf("secret value exceeds %d bytes", managedSecretMaxBytes)
	}
	value := string(raw)
	if !utf8.ValidString(value) {
		return "", errors.New("secret value must be valid UTF-8")
	}
	return value, nil
}

func runManagedSecretList(args []string) error {
	fs := flag.NewFlagSet("gestaltd secrets list", flag.ContinueOnError)
	fs.Usage = func() { printSimpleUsage(fs.Output(), "gestaltd secrets list") }
	if err := parseInterspersed(fs, args); err != nil {
		return err
	}
	if positionals := fs.Args(); len(positionals) > 0 {
		return fmt.Errorf("unexpected arguments: %s", strings.Join(positionals, " "))
	}

	client, err := newManagedSecretClient()
	if err != nil {
		return err
	}
	var all []managedSecretSummary
	if err := client.get(context.Background(), managedSecretsAdminPath, &all); err != nil {
		return fmt.Errorf("failed to list managed secrets: %w", err)
	}
	printManagedSecretRows(os.Stdout, all)
	return nil
}

func runManagedSecretDescribe(args []string) error {
	fs := flag.NewFlagSet("gestaltd secrets describe", flag.ContinueOnError)
	fs.Usage = func() { printSimpleUsage(fs.Output(), "gestaltd secrets describe NAME") }
	if err := parseInterspersed(fs, args); err != nil {
		return err
	}
	positionals := fs.Args()
	if len(positionals) != 1 {
		return fmt.Errorf("secret name is required")
	}
	name := positionals[0]

	client, err := newManagedSecretClient()
	if err != nil {
		return err
	}
	var response managedSecretDetail
	if err := client.get(context.Background(), managedSecretsAdminPath+"/"+url.PathEscape(name), &response); err != nil {
		return fmt.Errorf("failed to describe managed secret %s: %w", name, err)
	}
	printManagedSecretDescribeResponse(os.Stdout, response)
	return nil
}

func printManagedSecretDescribeResponse(w io.Writer, response managedSecretDetail) {
	_, _ = fmt.Fprintf(w, "Name: %s\n", response.Name)
	_, _ = fmt.Fprintf(w, "Owner app: %s\n", response.OwnerApp)
	_, _ = fmt.Fprintf(w, "Scope: %s\n", response.Scope)
	if response.Description != "" {
		_, _ = fmt.Fprintf(w, "Description: %s\n", response.Description)
	}
	_, _ = fmt.Fprintf(w, "Current version: %d\n", response.CurrentVersion)
	_, _ = fmt.Fprintf(w, "Created at: %s\n", response.CreatedAt)
	_, _ = fmt.Fprintf(w, "Updated at: %s\n", response.UpdatedAt)
	if response.RetiredAt != nil {
		_, _ = fmt.Fprintf(w, "Retired at: %s\n", *response.RetiredAt)
	}
	printManagedSecretAudit(w, response.Audit)
}

func runManagedSecretRetire(args []string) error {
	fs := flag.NewFlagSet("gestaltd secrets retire", flag.ContinueOnError)
	fs.Usage = func() { printSimpleUsage(fs.Output(), "gestaltd secrets retire NAME --reason TEXT") }
	reason := fs.String("reason", "", "why the secret is being retired")
	if err := parseInterspersed(fs, args); err != nil {
		return err
	}
	positionals := fs.Args()
	if len(positionals) != 1 {
		return fmt.Errorf("secret name is required")
	}
	name := positionals[0]
	if strings.TrimSpace(*reason) == "" {
		return fmt.Errorf("--reason is required")
	}

	client, err := newManagedSecretClient()
	if err != nil {
		return err
	}
	path := fmt.Sprintf("%s/%s/retire", managedSecretsAdminPath, url.PathEscape(name))
	var response managedSecretSummary
	if err := client.post(context.Background(), path, map[string]any{"reason": *reason}, &response); err != nil {
		return fmt.Errorf("failed to retire managed secret %s: %w", name, err)
	}
	_, _ = fmt.Fprintf(os.Stdout, "Retired %s\n", name)
	return nil
}

func clientLoginSupported(baseURL string) (bool, error) {
	var payload struct {
		LoginSupported bool `json:"loginSupported"`
	}
	client := &managedSecretClient{BaseURL: strings.TrimRight(baseURL, "/")}
	if err := client.get(context.Background(), "/api/v1/auth/info", &payload); err != nil {
		return false, fmt.Errorf("failed to determine authentication mode: %w", err)
	}
	return payload.LoginSupported, nil
}

func newManagedSecretClient() (*managedSecretClient, error) {
	baseURL, err := config.ResolveGestaltCLIURL()
	if err != nil {
		return nil, err
	}
	token, err := config.ResolveGestaltCLIToken()
	if err != nil {
		return nil, err
	}
	if strings.TrimSpace(baseURL) == "" {
		return nil, errors.New("gestalt URL is required; set GESTALT_URL or run `gestalt init`")
	}
	if strings.TrimSpace(token) == "" {
		loginSupported, authErr := clientLoginSupported(baseURL)
		if authErr != nil {
			return nil, authErr
		}
		if loginSupported {
			return nil, errors.New("gestalt credentials are required; set GESTALT_API_KEY or run `gestalt auth login`")
		}
	}
	return &managedSecretClient{
		BaseURL: strings.TrimRight(strings.TrimSpace(baseURL), "/"),
		Token:   strings.TrimSpace(token),
	}, nil
}

func (c *managedSecretClient) do(ctx context.Context, method, path string, body any) ([]byte, error) {
	if c == nil {
		return nil, errors.New("managed secrets client is not configured")
	}
	var reader io.Reader
	if body != nil {
		encoded, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		reader = bytes.NewReader(encoded)
	}
	req, err := http.NewRequestWithContext(ctx, method, c.BaseURL+path, reader)
	if err != nil {
		return nil, err
	}
	if c.Token != "" {
		req.Header.Set("Authorization", "Bearer "+c.Token)
	}
	if reader != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	httpClient := c.HTTPClient
	if httpClient == nil {
		httpClient = &http.Client{Timeout: managedSecretHTTPTimeout}
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%s %s: %w", method, path, err)
	}
	defer func() { _ = resp.Body.Close() }()
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read %s response: %w", method, err)
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		message := strings.TrimSpace(string(respBody))
		var payload struct {
			Error string `json:"error"`
		}
		if json.Unmarshal(respBody, &payload) == nil && strings.TrimSpace(payload.Error) != "" {
			message = strings.TrimSpace(payload.Error)
		}
		if message == "" {
			message = http.StatusText(resp.StatusCode)
		}
		return nil, fmt.Errorf("managed secrets API returned %d: %s", resp.StatusCode, message)
	}
	return respBody, nil
}

func (c *managedSecretClient) get(ctx context.Context, path string, out any) error {
	body, err := c.do(ctx, http.MethodGet, path, nil)
	if err != nil {
		return err
	}
	return decodeManagedSecretResponse(body, out)
}

func (c *managedSecretClient) post(ctx context.Context, path string, requestBody any, out any) error {
	body, err := c.do(ctx, http.MethodPost, path, requestBody)
	if err != nil {
		return err
	}
	return decodeManagedSecretResponse(body, out)
}

func decodeManagedSecretResponse(body []byte, out any) error {
	if out == nil {
		return nil
	}
	if len(bytes.TrimSpace(body)) == 0 {
		return nil
	}
	if err := json.Unmarshal(body, out); err != nil {
		return fmt.Errorf("decode response: %w", err)
	}
	return nil
}

type managedSecretSummary struct {
	Name           string  `json:"name"`
	OwnerApp       string  `json:"ownerApp"`
	Scope          string  `json:"scope"`
	Description    string  `json:"description"`
	CurrentVersion int64   `json:"currentVersion"`
	CreatedAt      string  `json:"createdAt"`
	UpdatedAt      string  `json:"updatedAt"`
	RetiredAt      *string `json:"retiredAt,omitempty"`
}

type managedSecretDetail struct {
	managedSecretSummary
	Audit []managedSecretAuditRow `json:"audit"`
}

type managedSecretAuditRow struct {
	Action    string `json:"action"`
	Version   int64  `json:"version"`
	Actor     string `json:"actor"`
	CreatedAt string `json:"createdAt"`
}

type managedSecretVersionSummary struct {
	Version   int64  `json:"version"`
	CreatedAt string `json:"createdAt"`
}

type managedSecretWriteResponse struct {
	Secret          managedSecretSummary        `json:"secret"`
	Version         managedSecretVersionSummary `json:"version"`
	GeneratedValue  string                      `json:"generatedValue,omitempty"`
	RolloutRequired bool                        `json:"rolloutRequired"`
}

func printManagedSecretRows(w io.Writer, rows []managedSecretSummary) {
	if len(rows) == 0 {
		_, _ = fmt.Fprintln(w, "No managed secrets")
		return
	}
	_, _ = fmt.Fprintf(w, "%-40s %-30s %-10s %8s\n", "NAME", "OWNER APP", "SCOPE", "VERSION")
	for i := range rows {
		_, _ = fmt.Fprintf(w, "%-40s %-30s %-10s %8d\n", rows[i].Name, rows[i].OwnerApp, rows[i].Scope, rows[i].CurrentVersion)
	}
}

func printManagedSecretAudit(w io.Writer, rows []managedSecretAuditRow) {
	if len(rows) == 0 {
		return
	}
	_, _ = fmt.Fprintln(w, "Audit:")
	for i := range rows {
		_, _ = fmt.Fprintf(w, "  %s v%d by %s: %s\n", rows[i].CreatedAt, rows[i].Version, rows[i].Actor, rows[i].Action)
	}
}

func printSimpleUsage(w io.Writer, usage string) {
	writeUsageLine(w, "Usage:")
	writeUsageLine(w, "  "+usage)
}

// readManagedSecretPassword reads a hidden value from a terminal.
func readManagedSecretPassword(label string) (string, error) {
	fmt.Fprint(os.Stderr, label)
	value, err := term.ReadPassword(int(os.Stdin.Fd()))
	fmt.Fprintln(os.Stderr)
	if err != nil {
		return "", err
	}
	return string(value), nil
}

File: gestalt/cli/src/commands/secrets.rs
use anyhow::{Context, Result, bail};
use std::io::{IsTerminal, Read};
use std::path::PathBuf;

use crate::api::ApiClient;
use crate::output::{self, Format};

const MANAGED_SECRETS_PATH: &str = "/admin/api/v1/managed-secrets";
const MANAGED_SECRET_MAX_BYTES: usize = 64 * 1024;

pub fn dispatch(client: &ApiClient, command: SecretsCommands, format: Format) -> Result<()> {
    match command {
        SecretsCommands::Create(args) => create_or_rotate(client, args, format),
        SecretsCommands::Rotate(args) => create_or_rotate(client, args, format),
        SecretsCommands::List { owner_app } => list(client, owner_app.as_deref(), format),
        SecretsCommands::Describe { name } => describe(client, &name, format),
        SecretsCommands::Preflight => preflight(client, format),
        SecretsCommands::Retire { name, reason } => retire(client, &name, &reason, format),
    }
}

fn create_or_rotate(
    client: &ApiClient,
    args: ManagedSecretWriteArgs,
    format: Format,
) -> Result<()> {
    let value = read_value(&args)?;
    let generate = args.generate;
    let mut body = serde_json::json!({
        "name": args.name,
        "ownerApp": args.owner_app.unwrap_or_default(),
        "scope": args.scope.unwrap_or_else(|| "app".to_string()),
        "description": args.description.unwrap_or_default(),
        "reason": args.reason,
        "source": "cli",
        "requestId": args.request_id.unwrap_or_default(),
    });
    if generate {
        body["generate"] = serde_json::json!(true);
    } else {
        body["value"] = serde_json::json!(value);
    }

    let response = client
        .post(MANAGED_SECRETS_PATH, &body)
        .context("failed to write managed secret")?;
    if format == Format::Json {
        output::print_json(&response);
        return Ok(());
    }

    if let Some(generated) = response.get("generatedValue").and_then(|v| v.as_str())
        && !generated.is_empty()
    {
        println!("Generated value (shown once):");
        println!("{generated}");
    }
    let secret = &response["secret"];
    let version = &response["version"];
    println!("Secret: {}", text(secret, "name"));
    if let Some(owner) = secret.get("ownerApp").and_then(|v| v.as_str())
        && !owner.is_empty()
    {
        println!("Owner app: {owner}");
    }
    println!("Version: {}", number(version, "version"));
    println!("Created at: {}", text(version, "createdAt"));
    if response
        .get("rolloutRequired")
        .and_then(|v| v.as_bool())
        .unwrap_or(false)
    {
        println!("Rollout required: yes");
        println!(
            "Gestalt resolves this secret during startup; roll out the runtime to load the new version."
        );
    } else {
        println!("Rollout required: no");
    }
    Ok(())
}

fn list(client: &ApiClient, owner_app: Option<&str>, format: Format) -> Result<()> {
    let response = client
        .get(MANAGED_SECRETS_PATH)
        .context("failed to list managed secrets")?;
    if format == Format::Json {
        output::print_json(&response);
        return Ok(());
    }
    let Some(rows) = response.as_array() else {
        bail!("managed secret list response was not an array");
    };
    let headers = ["Name", "Owner app", "Scope", "Version", "Updated"];
    let mut table = Vec::new();
    for row in rows {
        if let Some(owner) = owner_app
            && row.get("ownerApp").and_then(|v| v.as_str()) != Some(owner)
        {
            continue;
        }
        table.push(vec![
            text(row, "name"),
            text(row, "ownerApp"),
            text(row, "scope"),
            row.get("currentVersion")
                .and_then(|v| v.as_i64())
                .map(|v| v.to_string())
                .unwrap_or_else(|| "-".to_string()),
            text(row, "updatedAt"),
        ]);
    }
    output::print_table(&headers, &table);
    Ok(())
}

fn describe(client: &ApiClient, name: &str, format: Format) -> Result<()> {
    let encoded = crate::api::encode_path_segment(name);
    let response = client
        .get(&format!("{MANAGED_SECRETS_PATH}/{encoded}"))
        .with_context(|| format!("failed to describe managed secret {name}"))?;
    if format == Format::Json {
        output::print_json(&response);
        return Ok(());
    }
    println!("Name: {}", text(&response, "name"));
    println!("Owner app: {}", text(&response, "ownerApp"));
    println!("Scope: {}", text(&response, "scope"));
    if let Some(description) = nonempty(text(&response, "description")) {
        println!("Description: {description}");
    }
    println!(
        "Current version: {}",
        response
            .get("currentVersion")
            .and_then(|v| v.as_i64())
            .unwrap_or_default()
    );
    println!("Created at: {}", text(&response, "createdAt"));
    println!("Updated at: {}", text(&response, "updatedAt"));
    if let Some(retired) = response.get("retiredAt").and_then(|v| v.as_str()) {
        println!("Retired at: {retired}");
    }
    if let Some(audit) = response.get("audit").and_then(|v| v.as_array()) {
        println!("Audit:");
        for row in audit {
            println!(
                "  {} v{} by {}: {}",
                text(row, "createdAt"),
                row.get("version")
                    .and_then(|v| v.as_i64())
                    .unwrap_or_default(),
                text(row, "actor"),
                text(row, "action"),
            );
        }
    }
    Ok(())
}

fn preflight(client: &ApiClient, format: Format) -> Result<()> {
    // The CLI accepts references on stdin, matching the deploy preflight API.
    let mut raw = String::new();
    std::io::stdin()
        .read_to_string(&mut raw)
        .context("failed to read preflight references from stdin")?;
    let references: serde_json::Value =
        serde_json::from_str(&raw).context("failed to parse preflight references")?;
    let response = client
        .post(&format!("{MANAGED_SECRETS_PATH}/preflight"), &references)
        .context("failed to run managed secret preflight")?;
    if format == Format::Json {
        output::print_json(&response);
        return Ok(());
    }
    let ok = response
        .get("ok")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);
    println!("Status: {}", if ok { "PASS" } else { "FAIL" });
    if let Some(missing) = response.get("missing").and_then(|v| v.as_array()) {
        for row in missing {
            println!("Missing: {}", text(row, "name"));
            if let Some(app) = nonempty(text(row, "app")) {
                println!("  App: {app}");
            }
            if let Some(field) = nonempty(text(row, "field")) {
                println!("  Field: {field}");
            }
        }
    }
    if let Some(unreferenced) = response.get("unreferenced").and_then(|v| v.as_array()) {
        for row in unreferenced {
            println!("Unreferenced: {}", text(row, "name"));
        }
    }
    if !ok {
        std::process::exit(1);
    }
    Ok(())
}

fn retire(client: &ApiClient, name: &str, reason: &str, format: Format) -> Result<()> {
    let encoded = crate::api::encode_path_segment(name);
    let response = client
        .post(
            &format!("{MANAGED_SECRETS_PATH}/{encoded}/retire"),
            &serde_json::json!({ "reason": reason }),
        )
        .with_context(|| format!("failed to retire managed secret {name}"))?;
    if format == Format::Json {
        output::print_json(&response);
    } else {
        println!("Retired {name}");
    }
    Ok(())
}

fn read_value(args: &ManagedSecretWriteArgs) -> Result<String> {
    if args.generate {
        if args.value_file.is_some() {
            bail!("--generate and --file are mutually exclusive");
        }
        return Ok(String::new());
    }
    let raw: Vec<u8> = match args.value_file.as_ref() {
        Some(path) => {
            std::fs::read(path).with_context(|| format!("failed to read {}", path.display()))?
        }
        None => {
            if std::io::stdin().is_terminal() {
                let value = prompt_hidden_value().context("failed to read secret value")?;
                return Ok(value);
            }
            let mut buffer = Vec::new();
            std::io::stdin()
                .read_to_end(&mut buffer)
                .context("failed to read secret value from stdin")?;
            buffer
        }
    };
    if raw.len() > MANAGED_SECRET_MAX_BYTES {
        bail!("secret value exceeds 64 KiB");
    }
    let value = String::from_utf8(raw).context("secret value must be valid UTF-8")?;
    if value.is_empty() {
        bail!("secret value is required");
    }
    Ok(value)
}

fn prompt_hidden_value() -> Result<String> {
    let first = rpassword::prompt_password("Secret value: ")?;
    let second = rpassword::prompt_password("Confirm secret value: ")?;
    if first != second {
        bail!("secret values do not match");
    }
    if first.is_empty() {
        bail!("secret value is required");
    }
    Ok(first)
}

fn text(value: &serde_json::Value, key: &str) -> String {
    value
        .get(key)
        .and_then(|v| v.as_str())
        .map(str::to_string)
        .unwrap_or_else(|| "-".to_string())
}

fn number(value: &serde_json::Value, key: &str) -> i64 {
    value.get(key).and_then(|v| v.as_i64()).unwrap_or_default()
}

fn nonempty(value: String) -> Option<String> {
    if value.trim().is_empty() {
        None
    } else {
        Some(value)
    }
}

#[derive(Debug, Clone, clap::Args)]
pub struct ManagedSecretWriteArgs {
    /// Logical secret name. This name is also bound into the ciphertext as additional authenticated data.
    pub name: String,

    /// App that owns this secret.
    #[arg(long = "owner-app")]
    pub owner_app: Option<String>,

    /// Secret scope: app or shared.
    #[arg(long)]
    pub scope: Option<String>,

    /// Human-readable description.
    #[arg(long)]
    pub description: Option<String>,

    /// Why this value is being written.
    #[arg(long)]
    pub reason: String,

    /// Stable request identifier for automation retries.
    #[arg(long = "request-id")]
    pub request_id: Option<String>,

    /// Read exact value bytes from a file instead of stdin or a prompt.
    #[arg(long = "file")]
    pub value_file: Option<PathBuf>,

    /// Generate a 48-byte base64url secret and print it once.
    #[arg(long, default_value_t = false)]
    pub generate: bool,
}

#[derive(Debug, Clone, clap::Subcommand)]
pub enum SecretsCommands {
    /// Create a new managed secret version. Fails if the name already exists.
    Create(ManagedSecretWriteArgs),
    /// Rotate an existing managed secret by creating a new version.
    Rotate(ManagedSecretWriteArgs),
    /// List managed secret metadata. Values are never returned.
    List {
        /// Filter by owning app.
        #[arg(long = "owner-app")]
        owner_app: Option<String>,
    },
    /// Show one secret's metadata and recent audit history.
    Describe {
        /// Logical secret name.
        name: String,
    },
    /// Check configured references against stored secrets. Reads JSON references from stdin.
    Preflight,
    /// Mark a secret retired while retaining history.
    Retire {
        /// Logical secret name.
        name: String,
        /// Why the secret is being retired.
        #[arg(long)]
        reason: String,
    },
}
